File: scip-java/src/main/scala/com/sourcegraph/scip_java/buildtools/BuildTool.scala
package com.sourcegraph.scip_java.buildtools

import java.nio.file.Files
import java.nio.file.Path

import com.sourcegraph.scip_java.commands.IndexCommand
import os.CommandResult

/**
 * A build tool such as Gradle, Maven or Bazel.
 */
abstract class BuildTool(val name: String, index: IndexCommand) {
  def isHidden: Boolean = false
  final def sourceroot: Path = index.workingDirectory
  def usedInCurrentDirectory(): Boolean
  def generateScip(): Int
}

object BuildTool {
  def all(index: IndexCommand): List[BuildTool] =
    // We don't support Bazel for auto-indexing, but if it's
    // detected, we should at least give a meaningful error message
    autoOrdered(index) :+ new BazelBuildTool(index)

  def autoOrdered(index: IndexCommand): List[BuildTool] = List(
    // The order in this list is important -
    // first detected build tool will be used in `auto` mode
    // Bazel is missing because it isn't supported by auto-indexing

    // first as it indicates user's intent to use SCIP auto-indexing
    new ScipBuildTool(index),
    // Maven first, then Gradle
    // To match the order indicated in IntelliJ Java developer survey 2022:
    // https://www.jetbrains.com/lp/devecosystem-2022/java/#which-build-systems-do-you-regularly-use-if-any-
    new MavenBuildTool(index),
    new GradleBuildTool(index)
  )
  def allNames: String = all(IndexCommand())
    .filterNot(_.isHidden)
    .map(_.name)
    .mkString(", ")

  def generateScipFromTargetroot(
      generateScipResult: CommandResult,
      targetroot: Path,
      index: IndexCommand
  ): Int = {
    if (!Files.isDirectory(targetroot)) {
      generateScipResult.exitCode
    } else if (index.app.reporter.hasErrors()) {
      index.app.reporter.exitCode()
    } else if (generateScipResult.exitCode != 0) {
      generateScipResult.exitCode
    } else {
      index
        .indexScip
        .copy(
          output = index.finalOutput,
          targetroot = List(targetroot),
          app = index.app
        )
        .run()
    }
  }
}

File: scip-java/src/main/kotlin/com/sourcegraph/scip_java/buildtools/BuildTool.kt
package com.sourcegraph.scip_java.buildtools

import com.sourcegraph.scip_java.commands.AggregateRunner
import com.sourcegraph.scip_java.commands.IndexCommand
import java.nio.file.Files
import java.nio.file.Path

/** A build tool such as Gradle, Maven or Bazel. */
abstract class BuildTool(val name: String, protected val index: IndexCommand) {
    open val isHidden: Boolean = false

    val sourceroot: Path
        get() = index.workingDirectory

    abstract fun usedInCurrentDirectory(): Boolean

    abstract fun generateScip(): Int

    companion object {
        fun all(index: IndexCommand): List<BuildTool> =
            // We don't support Bazel for auto-indexing, but if it's
            // detected, we should at least give a meaningful error message.
            autoOrdered(index) + BazelBuildTool(index)

        fun autoOrdered(index: IndexCommand): List<BuildTool> =
            // The order in this list is important - first detected build
            // tool will be used in `auto` mode. Bazel is missing because
            // it isn't supported by auto-indexing.
            listOf(
                ScipBuildTool(index),
                MavenBuildTool(index),
                GradleBuildTool(index),
            )

        fun allNames(): String =
            all(IndexCommand()).filterNot { it.isHidden }.joinToString(", ") { it.name }

        /**
         * After the wrapped build tool finished invoking `javac`/`kotlinc`,
         * convert the resulting SCIP targetroot into a SCIP index.
         */
        fun generateScipFromTargetroot(
            generateScipResult: ProcessResult,
            targetroot: Path,
            index: IndexCommand,
        ): Int {
            if (!Files.isDirectory(targetroot)) return generateScipResult.exitCode
            if (index.app.reporter.hasErrors()) return index.app.reporter.exitCode()
            if (generateScipResult.exitCode != 0) return generateScipResult.exitCode
            return AggregateRunner.run(
                output = index.finalOutput,
                targetroots = listOf(targetroot),
                app = index.app,
                parallel = index.aggregateParallel,
                emitInverseRelationships = index.aggregateEmitInverseRelationships,
                allowEmptyIndex = index.aggregateAllowEmptyIndex,
                allowExportingGlobalSymbolsFromDirectoryEntries =
                    index.aggregateAllowExportingGlobalSymbolsFromDirectoryEntries,
            )
        }
    }
}
